util: Implement `Paginator` class as interface to access a list by pages

xdustinface opened the pull request:
Preparation for a follow up PR to provide paginated plot RPC endpoints. Im fine waiting with the merge until that other PR is ready, just opening this to already get it reviewed.

xdustinface committed:
util: Implement `Paginator` class as interface to access a list by pages

altendky commented on chia/util/paginator.py:
```diff
@@ -0,0 +1,37 @@
+    _source: Sequence[object]
+    _page_size: int
+
+    page_sizes: Tuple[int, ...] = (5, 10, 25, 50, 100)
```
I don't think this restriction improves the class.

xdustinface replied:
See [5d1e2a1](https://github.com/Chia-Network/chia-blockchain/pull/11247/commits/5d1e2a1a6bd8b1f3b1adfb6252e45ef205c6597e)

altendky commented on chia/util/paginator.py:
```diff
@@ -0,0 +1,37 @@
+    def page_count(self) -> int:
+        return max(1, ceil(len(self._source) / self._page_size))
+
+    def get_page(self, page: int) -> Sequence[object]:
```
Do we really want 1-based paging instead of 0-based?  I don't know what is normal for HTTP APIs.

xdustinface replied:
I don't know either. But i changed it to 0-based now because i was also not super convinced about the decision, see [a18ba58](https://github.com/Chia-Network/chia-blockchain/pull/11247/commits/a18ba589144c44954a8c6e6a56908c481f8ca944)

altendky reviewed (dismissed):
Can the looping tests be switched to parametrization to get individual results?  The others I would make separate tests for each assert as well so all tests happen and are available as results.

xdustinface committed:
Be less restrictive about page sizes and refactor tests

xdustinface committed:
Make the pages based of 0 instead of 1 and some more test refactoring

xdustinface committed:
More tests

xdustinface committed:
Adjust workflows after rebase

xdustinface commented:
@altendky See last few commits, improved tests (kind of mixed in commits with the other two changes), addressed the comments and rebased to get the correct workflows.

altendky commented on chia/util/paginator.py:
```diff
@@ -0,0 +1,46 @@
+        return max(1, ceil(len(self._source) / self._page_size))
+
+    def get_page(self, page: int) -> Sequence[object]:
+        if page < 0 or page > self.page_count() - 1:
```
```suggestion
        if 0 <= page < self.page_count():
```

xdustinface replied:
You wanna raise for good values?

altendky replied:
uh...  right...  i like 'em in order.  (remember when you had an approval and gave me more chances to complain?  `:]`)  but really, whatever here.

```suggestion
        if page < 0 or self.page_count() <= page:
```

xdustinface replied:
I almost like it, but what about [04d9b60](https://github.com/Chia-Network/chia-blockchain/pull/11247/commits/04d9b605c46d1f3dbd328f699a47728a0cbe249d) 😄 (swap the order)

altendky commented on tests/util/test_paginator.py:
```diff
@@ -0,0 +1,75 @@
+    ],
+)
+def test_get_page_valid(length: int, page: int, page_size: int, expected_data: List[int]) -> None:
+    assert Paginator(list(range(0, length)), page_size).get_page(page) == expected_data
```
`:]` maybe?

```suggestion
    assert Paginator(range(length), page_size).get_page(page) == expected_data
```

xdustinface replied:
`expected_data` is no range though :)

xdustinface committed:
Introduce `Paginator.create`

xdustinface committed:
`<=` instead of `- 1`
